File: frontend/src/ml/preprocessor.ts
import {
  ExtractedFrameLandmarks,
  FEATURE_DIMENSIONS,
  NormalizedLandmark,
  PreprocessedSequence,
  SEQUENCE_CONFIG,
} from './types';

export class LandmarkPreprocessor {
  /**
   * Preprocesses a single frame of raw MediaPipe Holistic landmarks into a 1629-dimensional Float32Array feature vector.
   */
  public static preprocessFrame(frame: ExtractedFrameLandmarks): Float32Array {
    const vector = new Float32Array(FEATURE_DIMENSIONS.TOTAL_PER_FRAME);
    let offset = 0;

    // 1. Pose Landmarks (99 values)
    if (frame.poseLandmarks && frame.poseLandmarks.length >= 33) {
      // Use midpoint of shoulders (index 11 and 12) as origin for pose normalization
      const leftShoulder = frame.poseLandmarks[11];
      const rightShoulder = frame.poseLandmarks[12];
      const originX = (leftShoulder.x + rightShoulder.x) / 2;
      const originY = (leftShoulder.y + rightShoulder.y) / 2;
      const originZ = (leftShoulder.z + rightShoulder.z) / 2;

      for (let i = 0; i < 33; i++) {
        const lm = frame.poseLandmarks[i];
        vector[offset++] = lm.x - originX;
        vector[offset++] = lm.y - originY;
        vector[offset++] = lm.z - originZ;
      }
    } else {
      offset += FEATURE_DIMENSIONS.POSE_LANDMARKS; // Zero-fill
    }

    // 2. Left Hand Landmarks (63 values)
    if (frame.leftHandLandmarks && frame.leftHandLandmarks.length >= 21) {
      // Use wrist (index 0) as origin for left hand normalization
      const wrist = frame.leftHandLandmarks[0];
      for (let i = 0; i < 21; i++) {
        const lm = frame.leftHandLandmarks[i];
        vector[offset++] = lm.x - wrist.x;
        vector[offset++] = lm.y - wrist.y;
        vector[offset++] = lm.z - wrist.z;
      }
    } else {
      offset += FEATURE_DIMENSIONS.LEFT_HAND_LANDMARKS; // Zero-fill
    }

    // 3. Right Hand Landmarks (63 values)
    if (frame.rightHandLandmarks && frame.rightHandLandmarks.length >= 21) {
      // Use wrist (index 0) as origin for right hand normalization
      const wrist = frame.rightHandLandmarks[0];
      for (let i = 0; i < 21; i++) {
        const lm = frame.rightHandLandmarks[i];
        vector[offset++] = lm.x - wrist.x;
        vector[offset++] = lm.y - wrist.y;
        vector[offset++] = lm.z - wrist.z;
      }
    } else {
      offset += FEATURE_DIMENSIONS.RIGHT_HAND_LANDMARKS; // Zero-fill
    }

    // 4. Face Landmarks (1404 values)
    if (frame.faceLandmarks && frame.faceLandmarks.length >= 468) {
      // Use nose tip (index 1) as origin for face normalization
      const noseTip = frame.faceLandmarks[1];
      for (let i = 0; i < 468; i++) {
        const lm = frame.faceLandmarks[i];
        vector[offset++] = lm.x - noseTip.x;
        vector[offset++] = lm.y - noseTip.y;
        vector[offset++] = lm.z - noseTip.z;
      }
    } else {
      offset += FEATURE_DIMENSIONS.FACE_LANDMARKS; // Zero-fill
    }

    return vector;
  }

  /**
   * Preprocesses a window of frames (default 30) into a PreprocessedSequence containing a 48,870 float buffer.
   */
  public static preprocessSequence(sequence: ExtractedFrameLandmarks[]): PreprocessedSequence {
    const sequenceLength = SEQUENCE_CONFIG.SEQUENCE_LENGTH;
    const featureDim = SEQUENCE_CONFIG.FEATURE_DIM;
    const buffer = new Float32Array(SEQUENCE_CONFIG.TOTAL_TENSOR_SIZE);

    // Fill sequence buffer; pad with zero-vectors if sequence length is less than expected window
    for (let i = 0; i < sequenceLength; i++) {
      if (i < sequence.length) {
        const frameVector = this.preprocessFrame(sequence[i]);
        buffer.set(frameVector, i * featureDim);
      }
    }

    return {
      timestamp: Date.now(),
      sequenceLength,
      featureDim,
      data: buffer,
    };
  }
}

File: ml/preprocessing/preprocessor.py
"""
SignBridge AI - Landmark Preprocessor Module (Python)
Milestone M2.1: Dataset & Data Pipeline Foundation
Strictly preserves M1.2 specification: 1629 features per frame, 30-frame sequence window.
"""

from typing import Dict, List, Optional, Any
import numpy as np
from ml.config import FEATURE_DIM, SEQUENCE_LENGTH, TOTAL_TENSOR_SIZE


class PythonLandmarkPreprocessor:

    @staticmethod
    def preprocess_frame(frame_landmarks: Dict[str, Any]) -> np.ndarray:
        """
        Preprocesses a single frame of raw MediaPipe Holistic landmarks into a 1,629-dimensional NumPy float32 vector.
        
        frame_landmarks expects dict structure:
          {
            "poseLandmarks": [{"x": float, "y": float, "z": float}, ...], # 33 points
            "leftHandLandmarks": [{"x": float, "y": float, "z": float}, ...], # 21 points
            "rightHandLandmarks": [{"x": float, "y": float, "z": float}, ...], # 21 points
            "faceLandmarks": [{"x": float, "y": float, "z": float}, ...] # 468 points
          }
        """
        vector = np.zeros(FEATURE_DIM, dtype=np.float32)
        offset = 0

        # 1. Pose Landmarks (99 values)
        pose_lms = frame_landmarks.get("poseLandmarks")
        if pose_lms and len(pose_lms) >= 33:
            left_shoulder = pose_lms[11]
            right_shoulder = pose_lms[12]
            origin_x = (left_shoulder["x"] + right_shoulder["x"]) / 2.0
            origin_y = (left_shoulder["y"] + right_shoulder["y"]) / 2.0
            origin_z = (left_shoulder["z"] + right_shoulder["z"]) / 2.0

            for i in range(33):
                lm = pose_lms[i]
                vector[offset] = lm["x"] - origin_x
                vector[offset + 1] = lm["y"] - origin_y
                vector[offset + 2] = lm["z"] - origin_z
                offset += 3
        else:
            offset += 99

        # 2. Left Hand Landmarks (63 values)
        lh_lms = frame_landmarks.get("leftHandLandmarks")
        if lh_lms and len(lh_lms) >= 21:
            wrist = lh_lms[0]
            for i in range(21):
                lm = lh_lms[i]
                vector[offset] = lm["x"] - wrist["x"]
                vector[offset + 1] = lm["y"] - wrist["y"]
                vector[offset + 2] = lm["z"] - wrist["z"]
                offset += 3
        else:
            offset += 63

        # 3. Right Hand Landmarks (63 values)
        rh_lms = frame_landmarks.get("rightHandLandmarks")
        if rh_lms and len(rh_lms) >= 21:
            wrist = rh_lms[0]
            for i in range(21):
                lm = rh_lms[i]
                vector[offset] = lm["x"] - wrist["x"]
                vector[offset + 1] = lm["y"] - wrist["y"]
                vector[offset + 2] = lm["z"] - wrist["z"]
                offset += 3
        else:
            offset += 63

        # 4. Face Landmarks (1404 values)
        face_lms = frame_landmarks.get("faceLandmarks")
        if face_lms and len(face_lms) >= 468:
            nose_tip = face_lms[1]
            for i in range(468):
                lm = face_lms[i]
                vector[offset] = lm["x"] - nose_tip["x"]
                vector[offset + 1] = lm["y"] - nose_tip["y"]
                vector[offset + 2] = lm["z"] - nose_tip["z"]
                offset += 3
        else:
            offset += 1404

        return vector

    @classmethod
    def preprocess_sequence(cls, sequence_frames: List[Dict[str, Any]]) -> np.ndarray:
        """
        Preprocesses a list of frame landmark dicts into a fixed (30, 1629) float32 NumPy matrix.
        Pads or truncates to exactly SEQUENCE_LENGTH (30) frames.
        """
        matrix = np.zeros((SEQUENCE_LENGTH, FEATURE_DIM), dtype=np.float32)

        for i in range(min(len(sequence_frames), SEQUENCE_LENGTH)):
            matrix[i] = cls.preprocess_frame(sequence_frames[i])

        return matrix
